eth/backend.go

```diff
@@ -559,6 +576,75 @@ func (self *Ethereum) syncAccounts(tx *types.Transaction) {
 	}
 }
 
+func (self *Ethereum) StartAutoDAG() {
+	go func() {
+		glog.V(logger.Info).Infof("Automatic pregeneration of ethash DAG ON (ethash dir: %s)", self.AutoDAGDir)
+		var nextEpoch uint64
+		ticker := time.NewTicker(autoDAGcheckInterval)
+		self.autodagquit = make(chan bool)
+		for {
+			select {
+			case <-ticker.C:
+				currentBlock := self.ChainManager().CurrentBlock().NumberU64()
+				thisEpoch := currentBlock / epochLength
+				if nextEpoch <= thisEpoch {
+					if currentBlock%epochLength > autoDAGepochHeight {
+						if thisEpoch > 0 {
+							previousDag, previousDagFull := dagFiles(thisEpoch - 1)
+							os.Remove(filepath.Join(self.AutoDAGDir, previousDag))
+							os.Remove(filepath.Join(self.AutoDAGDir, previousDagFull))
+							glog.V(logger.Info).Infof("removed DAG for epoch %d (%s)", thisEpoch-1, previousDag)
+						}
+						nextEpoch = thisEpoch + 1
+						dag, _ := dagFiles(nextEpoch)
+						if _, err := os.Stat(dag); os.IsNotExist(err) {
+							glog.V(logger.Info).Infof("Pregenerating DAG for epoch %d (%s)", nextEpoch, dag)
+							err := self.MakeDAG(nextEpoch*epochLength, self.AutoDAGDir)
+							if err != nil {
+								glog.V(logger.Error).Infof("Error generating DAG for epoch %d (%s)", nextEpoch, dag)
+								return
+							}
+						} else {
+							glog.V(logger.Error).Infof("DAG for epoch %d (%s)", nextEpoch, dag)
+						}
+					}
+				}
+			case <-self.autodagquit:
+				return
+			}
+		}
+	}()
+}
+
+func dagFiles(epoch uint64) (string, string) {
+	seedHash, _ := ethash.GetSeedHash(epoch * epochLength)
+	dag := fmt.Sprintf("full-R%d-%x", ethashRevision, seedHash[:8])
+	return dag, "full-R" + dag
+}
+
+// stopAutoDAG stops automatic DAG pregeneration by quitting the loop
+func (self *Ethereum) StopAutoDAG() {
+	if self.autodagquit != nil {
+		close(self.autodagquit)
+		self.autodagquit = nil
+	}
+	glog.V(logger.Info).Infof("Automatic pregeneration of ethash DAG OFF (ethash dir: %s)", self.AutoDAGDir)
+}
+
+func (self *Ethereum) MakeDAG(blockNum uint64, dir string) (err error) {
+	dir = filepath.Clean(dir)
+	// seems to require a trailing slash
+	if !strings.HasSuffix(dir, "/") {
```

Comment:
Maybe better if directory separators are converted to `filepath.Separator` for OS compatibility? Same with line below